### PR TITLE
[Kotlin Cleanup] CustomStudyDialog

### PR DESCRIPTION
**Pull request by kilitr**
## Purpose / Description
Reviewing all `@KotlinCleanup` annotations from `CustomStudyDialog.kt` and refactoring them in order to improve code quality.

## Fixes
Related to #10489.

## Approach
9439fea5bed0a23fdffec6e24a87dcc1da2c81fe (Most annotations in the file are fixed here) should be an easy review and (in my belief) does not raise any questions.

cc397f26b264ae76bb62d2e53b20489543703fb7 (only about this bit: `@KotlinCleanup("Map dialogID to enum to make exhaustive")`) It's quite a lot and I'm not fully convinced, that it is an improvement.

## How Has This Been Tested?
local testing

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 1]
```diff
@@ -204,7 +198,7 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
                 when (dialogId) {
                     CUSTOM_STUDY_NEW -> {
                         AnkiDroidApp.getSharedPrefs(requireActivity()).edit().putInt("extendNew", n).apply()
-                        collection?.let {
+                        collection.let {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:201]: Can we remove the `let`

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 2]
```diff
@@ -214,7 +208,7 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
                     }
                     CUSTOM_STUDY_REV -> {
                         AnkiDroidApp.getSharedPrefs(requireActivity()).edit().putInt("extendRev", n).apply()
-                        collection?.let {
+                        collection.let {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:211]: can we remove the `let`

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 3]
```diff
@@ -374,18 +368,16 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
                     CUSTOM_STUDY_FORGOT, CUSTOM_STUDY_AHEAD, CUSTOM_STUDY_RANDOM,
                     CUSTOM_STUDY_PREVIEW, CUSTOM_STUDY_TAGS, DECK_OPTIONS
                 )
-            else -> {
-            }
+            else -> return intArrayOf()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: This should fail, we're producing a dialog with no list items. Null is fine for a return value until we resolve the other property: that the check should be exhaustive

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt [hunk 4]
```diff
@@ -99,7 +99,7 @@ class CustomStudyDialogTest : RobolectricTest() {
     @KotlinCleanup("Use kotlin based Mockito extensions")
     fun increaseNewCardLimitRegressionTest() {
         // #8338 - Regression Test
-        val args = CustomStudyDialog(ParametersUtils.whatever(), ParametersUtils.whatever())
+        val args = CustomStudyDialog(col, ParametersUtils.whatever())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt:102]: `any()` (from `import org.mockito.kotlin.any`) should work here

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 5]
```diff
@@ -258,8 +258,7 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
                             false
                         )
                     }
-                    else -> {
-                    }
+                    else -> {}
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:261]: This refactoring doesn't feel worthwhile if we're not getting rid of the `else`s

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 6]
```diff
@@ -334,76 +333,74 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
...
-                        intArrayOf(CUSTOM_STUDY_REV, DECK_OPTIONS, MORE_OPTIONS)
+                        intArrayOf(ContextMenuOption.STUDY_REV.value, ContextMenuOption.DECK_OPTIONS.value, ContextMenuOption.MORE_OPTIONS.value)
                     }
                 }
-            CONTEXT_MENU_EMPTY_SCHEDULE -> // Special custom study options to show when extending the daily study limits is not applicable
+            ContextMenuConfiguration.EMPTY_SCHEDULE -> // Special custom study options to show when extending the daily study limits is not applicable
                 return intArrayOf(
-                    CUSTOM_STUDY_FORGOT, CUSTOM_STUDY_AHEAD, CUSTOM_STUDY_RANDOM,
-                    CUSTOM_STUDY_PREVIEW, CUSTOM_STUDY_TAGS, DECK_OPTIONS
+                    ContextMenuOption.STUDY_FORGOT.value, ContextMenuOption.STUDY_AHEAD.value, ContextMenuOption.STUDY_RANDOM.value,
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:366]: Ideally import `ContextMenuOption` and `ContextMenuConfiguration` so this change has less visual noise

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 7]
```diff
@@ -334,76 +333,74 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
      * @param dialogId option to specify which tasks are shown in the list
      * @return the ids of which values to show
      */
-    @KotlinCleanup("Map dialogID to enum to make exhaustive")
-    private fun getListIds(dialogId: Int): IntArray {
+    private fun getListIds(dialogId: ContextMenuConfiguration): IntArray {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:336]: Can we have this return a `List<ContextMenuOption>`?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 8]
```diff
@@ -489,25 +486,50 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
...
 
-    companion object {
+    interface ContextMenuAttribute<T> where T : Enum<*> {
+        val value: Int
+        companion object {
+            fun fromInt(value: Int): ContextMenuAttribute<*> {
+                return when (value) {
+                    in ContextMenuConfiguration.range() -> ContextMenuConfiguration.fromInt(value)
+                    in ContextMenuOption.range() -> ContextMenuOption.fromInt(value)
+                    else -> throw IllegalArgumentException("Unknown value for a custom study dialog context menu attribute was passed.")
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:496]: What would have happened in the previous case?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 9]
```diff
@@ -287,16 +286,16 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
...
-            keyValueMap[CUSTOM_STUDY_NEW] = res.getString(R.string.custom_study_increase_new_limit)
-            keyValueMap[CUSTOM_STUDY_REV] = res.getString(R.string.custom_study_increase_review_limit)
-            keyValueMap[CUSTOM_STUDY_FORGOT] = res.getString(R.string.custom_study_review_forgotten)
-            keyValueMap[CUSTOM_STUDY_AHEAD] = res.getString(R.string.custom_study_review_ahead)
-            keyValueMap[CUSTOM_STUDY_RANDOM] = res.getString(R.string.custom_study_random_selection)
-            keyValueMap[CUSTOM_STUDY_PREVIEW] = res.getString(R.string.custom_study_preview_new)
-            keyValueMap[CUSTOM_STUDY_TAGS] = res.getString(R.string.custom_study_limit_tags)
-            keyValueMap[DECK_OPTIONS] = res.getString(R.string.menu__deck_options)
-            keyValueMap[MORE_OPTIONS] = res.getString(R.string.more_options)
+            keyValueMap[ContextMenuConfiguration.STANDARD.value] = res.getString(R.string.custom_study)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:289]: Would this be better as a map from `ContextMenuOption` so we don't need `.value`?

File: AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt (hunk 4 above)
Comment (kilitr) [reply AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt:102]: this does not work as any() is a matcher and does not mock in any way.

however using mock() from org.mockito.kotlin would work if you prefer this.

Comment (david-allison) [reply AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt]: That works! I also added `strictMock()` if we want to ensure the mock isn't called.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt (hunk 9 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: Might be worthwhile to add a `KotlinCleanup` on this one - we can generify `getValuesFromKeys` very easily - I'll do so in a separate PR

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 10]
```diff
@@ -497,25 +483,50 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
...
 
+        companion object {
+            fun fromInt(value: Int): ContextMenuConfiguration = values().first { it.value == value }
+            fun range(): Range<Int> = Range.create(values().minOf { it.value }, values().maxOf { it.value })
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    enum class ContextMenuOption(override val value: Int) : ContextMenuAttribute<ContextMenuOption> {
         // Standard custom study options to show in the context menu
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:513]: convert to doccomment

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 11]
```diff
@@ -497,25 +483,50 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
...
+                    in ContextMenuConfiguration.range() -> ContextMenuConfiguration.fromInt(value)
+                    in ContextMenuOption.range() -> ContextMenuOption.fromInt(value)
+                    else -> throw IllegalArgumentException("Unknown value for a custom study dialog context menu attribute was passed.")
+                }
+            }
+        }
+    }
+
+    enum class ContextMenuConfiguration(override val value: Int) : ContextMenuAttribute<ContextMenuConfiguration> {
         // Different configurations for the context menu
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:500]: Convert to doccomment

**Review by david-allison — COMMENTED** [review]
Looks awesome!

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt (hunk 8 above)
Comment (kilitr) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:505]: searching for an answer to your question, i found - this method is unnecessary.

**Review by david-allison — APPROVED** [review]
Awesome, thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 12]
```diff
@@ -497,25 +483,52 @@ class CustomStudyDialog(@KotlinCleanup("Make collection non null") private val c
...
+        STUDY_AHEAD(103),
+        STUDY_RANDOM(104),
+        STUDY_PREVIEW(105),
+        STUDY_TAGS(106),
+        DECK_OPTIONS(107),
+        MORE_OPTIONS(108);
+
+        companion object {
+            fun fromInt(value: Int): ContextMenuOption = values().first { it.value == value }
+            fun range(): Range<Int> = Range.create(values().minOf { it.value }, values().maxOf { it.value })
```
Comment (kilitr) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:531]: i belief we also dont need `range` anymore... checking that real quick

Comment (kilitr) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: yes - removed

**Review by mikehardy — APPROVED** [review]
LGTM - pending ci

**Pull request merged**